File: linked list problems/C++/DLL.cpp
#include <iostream>

using namespace std;

struct Node
{
	int data;
	struct Node *next = NULL;
	struct Node *prev = NULL;
};


struct Node* create_dll()
{
	struct Node *head = NULL;
	bool done = false;

	cout<<"Please continue entering the data to the linkedlist and type done to stop it\n";

	while (!done){
		string input;
		cin>>input;
		
		if (input =="done")
			break;
		int data = stoi(input);

		struct Node *new_node = (struct Node*) malloc (sizeof(struct Node));
		new_node->data = data;
		new_node->next = NULL;
		new_node->prev = NULL;

		if (head==NULL)
			head = new_node;
		else
		{
			struct Node *temp = head;

			while(temp->next!=NULL)
				temp = temp->next;
			temp->next = new_node;
			new_node->prev = temp;	
		}

		cout<<"New node "<<data<<" is added to the linked list and type done to stop it"<<endl;

	}

	return head;

}

void printLinkedList(struct Node* head )
{
	struct Node* itr = head;
	struct Node* last = NULL;
	cout<<"-------------"<<endl;
	while(itr!=NULL)
	{
		last = itr;
		cout<<itr->data<<" ";
		itr = itr->next;
	}

	cout<<endl;

	while(last !=NULL)
	{
		cout<<last->data<<" ";
		last = last->prev;
	}

	cout<<endl;
	cout<<"-------------"<<endl;


}

struct Node* insert_pos(struct Node *head)
{
	int curr_pos = 1;
	struct Node *temp = head;
	cout<<"Please enter the data to be inserted"<<endl;
	string input;
	cin>>input;
	int data = stoi(input);

	cout<<"Please enter the index where you wanna insert the node"<<endl;
	cin>>input;

	int pos = stoi(input);

	struct Node *new_node = (struct Node*) malloc (sizeof(struct Node));
	new_node->next = NULL;
	new_node->prev = NULL;
	new_node->data = data;

	if (pos == 0)
	{
		new_node->next = head;
		head->prev = new_node;
		head = new_node;
		cout<<"The node has been inserted at "<<pos<<" index"<<endl;
		return head;
	}

	while(temp->next!=NULL)
	{
		if (curr_pos==pos)
		{
			new_node->next = temp->next;
			temp->next->prev = new_node;
			temp->next = new_node;
			new_node->prev = temp;

			cout<<"The node has been inserted at "<<pos<<" index"<<endl;
			return head;
		}
		temp = temp->next;
		curr_pos+=1;

		if (temp->next==NULL && curr_pos==pos)
		{
			temp->next = new_node;
			new_node->prev= temp;
			curr_pos+=1;
			cout<<"The node has been inserted at "<<pos<<" index"<<endl;
			return head;
		}
	}

	cout<<"index is out of range or invalid"<<endl;
	return head;
}

struct Node* delete_pos(struct Node *head)
{
	cout<<"Please enter the index where you wanna delete the node"<<endl;
	string input;
	cin>>input;

	int pos = stoi(input);
	int curr_pos=1;

	struct Node *temp = head;

	if (pos == 0)
	{
		head->next->prev = NULL;
		head = head->next;
		cout<<"The node at "<<pos<<" has been deleted"<<endl;
		return head;
	}

	while(temp->next!= NULL)
	{
		if (curr_pos==pos)
		{
			temp->next = temp->next->next;
			if (temp->next!=NULL)
				temp->next->prev = temp;
			cout<<"The node at "<<pos<<" has been deleted"<<endl;
			return head;
		}

		temp = temp->next;
		curr_pos+=1;

		if (temp->next==NULL && curr_pos==pos)
		{
			temp->prev->next = NULL;
			temp->prev= NULL;

			cout<<"The node at "<<pos<<" has been deleted"<<endl;
			return head;
		}
	}

	cout<<"index is out of range or invalid"<<endl;
	return head;

}

int main()
{
	struct Node *head = create_dll();
	printLinkedList(head);
	// struct Node *test = insert_pos(head);
	// printLinkedList(test);
	struct Node *test = delete_pos(head);
	printLinkedList(test);

	return 0;
}

File: linked list problems/Go/DLL.go
package main

import (
	"fmt"
	"os"
	"bufio"
	"strconv"
)

type Node struct {
	data int
	next *Node
	prev *Node

}


func create_DLL() *Node {

	scanner:= bufio.NewScanner(os.Stdin)
	var head* Node = nil
	
	fmt.Println("Please entet the data to create a DLL")

	for {
		scanner.Scan()
		text := scanner.Text()

		if text == "done" {
			break
		}

		data, err := strconv.Atoi(text)

		if err != nil {
			return head
		} 

		var new_node = Node{data,nil,nil}

		if head == nil {
			head = &new_node
		} else {

			temp := head

			for {

				if temp.next!=nil {
					temp = temp.next
				} else {
					break
				}
			}

			new_node.prev = temp
			temp.next = &new_node

		}

		fmt.Println("New node",data," is added to the linked list and type done to stop it")

	}
	return head
}

func printLinkedlist(head* Node) {
	var temp* Node = head
	var last* Node = nil

	fmt.Println("------------")
	for {
		if temp!=nil {
			last = temp
			fmt.Print(temp.data," ")
			temp = temp.next
		} else {
			break
		}
	}
	fmt.Println()
	for {
		if last!=nil {
			fmt.Print(last.data," ")
			last = last.prev
		} else {
			break
		}
	}
	fmt.Println()
	fmt.Println("------------")

}

func insert_pos(head* Node) *Node{
	fmt.Println("Please enter the data to insert")
	scanner := bufio.NewScanner(os.Stdin)

	scanner.Scan()
	text := scanner.Text()
	data, err := strconv.Atoi(text)

	var new_node = Node{data,nil,nil}

	fmt.Println("Please enter the index where the node should be inserted")
	scanner.Scan()
	text_pos := scanner.Text()
	pos, err := strconv.Atoi(text_pos)

	if err!=nil {
		return head
	}
	curr_pos := 1

	if pos == 0 {
		new_node.next = head
		head.prev = &new_node
		head = &new_node
		fmt.Println("The node has been inserted successfully")
		return head
	}	
	var temp* Node = head

	for {
		if temp.next!=nil {

			if curr_pos ==pos {
				new_node.next = temp.next
				new_node.prev = temp
				temp.next.prev = &new_node
				temp.next = &new_node
				fmt.Println("The node has been inserted successfully")
				return head
			}
			temp = temp.next
			curr_pos+=1

			if temp.next==nil && curr_pos==pos {
				temp.next = &new_node
				new_node.prev = temp
				fmt.Println("The node has been inserted successfully")
				return head
			}
		} else {
			break
		}
	}

	fmt.Println("the index is out of range or invalid")
	return head

}

func delete_pos(head* Node) *Node{

	fmt.Println("Please enter the index where the node should be deleted")
	scanner := bufio.NewScanner(os.Stdin)

	scanner.Scan()
	text := scanner.Text()
	pos, err := strconv.Atoi(text)
	curr_pos:=1
	if err!=nil {
		return head
	}

	if pos == 0 {
		head = head.next
		head.prev = nil
		
		fmt.Println("The node has been deleted successfully")
		return head
	}

	var temp* Node = head



	for {
		if temp.next!=nil {

			if curr_pos == pos {
				temp.next = temp.next.next
				if temp.next!=nil {
					temp.next.prev = temp
				}
				fmt.Println("The node has been deleted successfully")
				return head
			}
			temp = temp.next
			curr_pos+=1

			if temp.next==nil && curr_pos==pos {
				temp.prev.next = nil
				temp.prev = nil
				fmt.Println("The node has been deleted successfully")
				return head
			}
		} else {
			break
		}
	}

	fmt.Println("the index is out of range or invalid")
	return head


}

func reverse(head* Node) *Node {
	var prev* Node = nil
	var curr* Node = head
	var next* Node = nil

	for {
		if curr!=nil {
			next = curr.next
			curr.next = curr.prev
			curr.prev = next
			prev = curr
			curr = next
		} else {
			head = prev
			return head
		}
	}
}

func main() {
	head := create_DLL()
	printLinkedlist(head)
	// test := insert_pos(head)
	// printLinkedlist(test)
	test := delete_pos(head)
	printLinkedlist(test)
	test1 := reverse(head)
	printLinkedlist(test1)
}